File: Breadth First Search/111. Minimum Depth of Binary Tree.cpp
/*
BFS
when we finish one level: count + 1
if the node has no left and right children. return the count.
*/



/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int minDepth(TreeNode* root) {
        if (root == nullptr)
            return 0;
        int result = 1;
        queue<TreeNode*> nodeQueue;
        nodeQueue.push(root);
        while (!nodeQueue.empty())
        {
            int size = nodeQueue.size();
            for (int i = 0; i < size; i++)
            {
                TreeNode* cur = nodeQueue.front();
                nodeQueue.pop();
                if (cur->left == nullptr && cur->right == nullptr)
                    return result;
                if (cur->left != nullptr)
                    nodeQueue.push(cur->left);
                if (cur->right != nullptr)
                    nodeQueue.push(cur->right);
            }
            result++;
        }
        return result;
        
    }
};

File: Python Solutions/111. Minimum Depth of Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
"""
Use divide and conquer.

What is the minimum depth if we already get the left branch and right branch depth:
    We just get the min(leftBranchDepth, rightBranchDepth) + 1
However, the derivation above has an assumption: both left branch and right branch exist.
If not, we just need to focus on the side that exists.
"""

class Solution:
    def minDepth(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0
        if root.left is None and root.right is None:
            return 1
        elif root.left and root.right:
            return min(self.minDepth(root.left), self.minDepth(root.right)) + 1
        elif root.left:
            return self.minDepth(root.left) + 1
        else:
            return self.minDepth(root.right) + 1
        
